pallets/bonds/src/lib.rs

```diff
@@ -0,0 +1,340 @@
+	#[pallet::getter(fn bonds)]
+	pub(super) type Bonds<T: Config> = StorageMap<_, Blake2_128Concat, T::AssetId, Bond<T>>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A bond asset was registered
+		BondTokenCreated {
+			issuer: T::AccountId,
+			asset_id: T::AssetId,
+			bond_asset_id: T::AssetId,
+			amount: T::Balance,
+			fee: T::Balance,
+		},
+		/// Bonds were redeemed
+		BondsRedeemed {
+			who: T::AccountId,
+			bond_id: T::AssetId,
+			amount: T::Balance,
+		},
+		/// Bonds were unlocked
+		BondsUnlocked { bond_id: T::AssetId },
+	}
+
+	#[pallet::error]
+	#[cfg_attr(test, derive(PartialEq, Eq))]
+	pub enum Error<T> {
+		/// Bond not registered
+		BondNotRegistered,
+		/// Bond is not mature
+		BondNotMature,
+		/// Maturity not long enough
+		InvalidMaturity,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Issue new bonds.
+		/// New asset id is registered and assigned to the bonds.
+		/// The number of bonds the issuer receives is 1:1 to the `amount` of the underlying asset
+		/// minus the protocol fee.
+		/// The bond asset is registered with the empty string for the asset name,
+		/// and with the same existential deposit as of the underlying asset.
+		/// Bonds can be redeemed 1:1 for the underlying asset once mature.
+		/// Protocol fee is applied to the amount, and transferred to `T::FeeReceiver`.
+		///
+		/// Parameters:
+		/// - `origin`: issuer of new bonds, needs to be `T::IssueOrigin`
+		/// - `asset_id`: underlying asset id
+		/// - `amount`: the amount of the underlying asset
+		/// - `maturity`: Unix time in milliseconds, when the bonds will be mature. Needs to be set
+		/// more than `T::MinMaturity` from now.
+		///
+		/// Emits `BondTokenCreated` event when successful.
+		///
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config>::WeightInfo::issue())]
+		pub fn issue(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			amount: T::Balance,
+			maturity: Moment,
+		) -> DispatchResult {
+			let who = T::IssueOrigin::ensure_origin(origin)?;
+
+			let time_diff = maturity
+				.checked_sub(T::TimestampProvider::now())
+				.ok_or(ArithmeticError::Overflow)?;
+			ensure!(time_diff >= T::MinMaturity::get(), Error::<T>::InvalidMaturity);
+
+			let ed = T::ExistentialDeposits::get(&asset_id);
+
+			// not covered in the tests. Create an asset with empty name should always work
+			let bond_asset_id = T::AssetRegistry::create_bond_asset(ed)?;
+
+			let fee = T::ProtocolFee::get().mul_ceil(amount); // TODO: check
+			let amount_without_fee = amount.checked_sub(&fee).ok_or(ArithmeticError::Overflow)?;
+			let pallet_account = Self::pallet_account_id();
+
+			T::Currency::transfer(asset_id, &who, &pallet_account, amount_without_fee)?;
+			T::Currency::transfer(asset_id, &who, &T::FeeReceiver::get(), fee)?;
+			T::Currency::deposit(bond_asset_id, &who, amount_without_fee)?;
+
+			Bonds::<T>::insert(
```

Comment:
not every issuance creates new token
you can reissue more of the same bonds, bond token is effectively tied to the pair of asset & maturity, therefore when you issue HDX with the same maturity there will be just more of the same tokens issued

Reply:
i guess you need to have storage `(asset, maturity) -> bondId` to be able to do this as well

Reply:
fixed